### PR TITLE
DM-14673 Label for marker disappears when it is moved to the edge and footprint translation is out of shape on HiPS

### PR DESCRIPTION
**Pull request by cwang2016**
This development fixes the following issues, 
- when the marker/footprint is moved towards the edge of the display area, the label for the marker/footprint disappears.
   fixing: showing the label when the marker/footprint is moved out of plot area and showing the label in alternate corner location in case the label is being moved out of plot area. 
- when the marker/footprint is relocated farther from the center or the target of the HiPS, the footprint  is shown out of shape as the image is zoomed. 
  fixing: instead of translating each region component on image domain as regular image, each component is re-rendered in world coordinate after relocating the footprint center first per original region description. 

test:
start image search 
add a marker or footprint overlay
move the marker/footprint to anywhere close to the plot border and view the text display

start HiPS search (no FOV entered)
add a footprint overlay
click (relocate) at a point close to the curved border of the HiPS
zoom the image to see more detail of the footprint by zooming and rotating the HIPS

### CODE REVIEW

**Comment by cwang2016** [comment]
Add some update to fix footprint rendering on HiPS in sync with the change of projection center and the zooming.

File: src/firefly/js/drawingLayers/FootprintTool.js [hunk 1]
```diff
@@ -342,6 +347,23 @@ function getLayerChanges(drawLayer, action) {
             if (!isEmpty(get(drawLayer, ['drawData', 'data']))) {
                 return attachToNewPlot(drawLayer, get(action.payload, ['plotIdAry', '0']));
             }
+            break;
+        case ImagePlotCntlr.CHANGE_CENTER_OF_PROJECTION:
+        case ImagePlotCntlr.ANY_REPLOT:
```
Comment (robyww) [inline src/firefly/js/drawingLayers/FootprintTool.js]: you could get a bunch of these events.  Have you check scrolling performance to make sure it is not compromised?

File: src/firefly/js/visualize/draw/MarkerFootprintObj.js [hunk 2]
```diff
@@ -1080,44 +1085,62 @@ export function drawMarkerObject(drawObjP, ctx, plot, def, vpPtM, onlyAddToPath)
...
+                const tLoc = textLocSeq[t % textLocSeq.length];
+
+                const textPt = makeTextLocationComposite(plot, tLoc, fontSize,
+                    objArea.width * plot.zoomFactor,
+                    objArea.height * plot.zoomFactor,
+                    objArea.centerPt);
+                if (textPt) {
+                    drawText(drawObj, ctx, plot, textPt, drawParams);
+                    if (drawObj.pointNotInDisplay) {
+                        drawObj.pointNotInDisplay = undefined;
```
Comment (robyww) [inline src/firefly/js/visualize/draw/MarkerFootprintObj.js]: drawObj appears to be from the redux store.  It should not be modified.

File: src/firefly/js/visualize/draw/ShapeDataObj.js [hunk 3]
```diff
@@ -719,6 +719,8 @@ export function drawText(drawObj, ctx, plot, inPt, drawParams) {
         );
 
         drawObj.textWorldLoc = plot.getImageCoords(makeDevicePt(x, y));
+    } else {
+        drawObj.pointNotInDisplay = true;  // note on failure
```
Comment (robyww) [inline src/firefly/js/visualize/draw/ShapeDataObj.js]: same thing: drawObj is defined in the redux store.

File: src/firefly/js/drawingLayers/FootprintTool.js [hunk 4]
```diff
@@ -342,6 +347,23 @@ function getLayerChanges(drawLayer, action) {
...
+            console.log('type '+action.type);
+            if (plotIdAry) {
+                plotIdAry.forEach((pId) => {
+                    if (isGoodPlot(pId)) {
+                        wptObj = get(dd, ['data', pId, 'pts', '0']);
+                        const cc = getCC(pId);
+
+                        if (isHiPS(cc) && wptObj && isPointInView(wptObj, cc)) {
```
Comment (robyww) [inline src/firefly/js/drawingLayers/FootprintTool.js]: should `isPointInView` be moved to `CsysConverter.js`?

**Pull request merged**